instance/models/server.py

```diff
@@ -282,7 +287,10 @@ def public_ip(self):
         if not self.openstack_id:
             return None
 
-        public_addr = openstack.get_server_public_address(self.os_server)
+        try:
+            public_addr = openstack.get_server_public_address(self.os_server)
+        except novaclient.exceptions.NotFound:
+            return None  # This server has been deleted.
```

Comment:
Yeah, right now the code does not really properly handle the situation where an external user has deleted the OpenStack VM. For now I've just made this so it won't break when that happens (as you experienced), but I do agree we should have a better way to handle it in the future. I will add that to the list of suggested next steps.